src/arithmetic/bigint.rs

```diff
@@ -512,13 +514,19 @@ pub fn elem_exp_consttime<M>(
 
     let num_limbs = m.limbs().len();
 
+    // This alignment was inherited from upstream. Assume that some of the
+    // assembly language functions require it for correctness and/or for
+    // performance.
     const ALIGNMENT: usize = 64;
-    assert_eq!(ALIGNMENT % LIMB_BYTES, 0);
-    let mut table = vec![0; ((TABLE_ENTRIES + 3) * num_limbs) + ALIGNMENT];
+    const _ALIGNMENT_IS_MULTIPLE_OF_LIMB_SIZE: () =
+        assert!(ALIGNMENT % mem::size_of::<Limb>() == 0);
+    const _ALIGNMENT_IS_MULTIPLE_OF_LIMB_ALIGNMENT: () =
+        assert!(ALIGNMENT % mem::align_of::<Limb>() == 0);
+    const EXTRA_FOR_ALIGNMENT: usize = (ALIGNMENT / mem::size_of::<Limb>()) - 1;
+    let mut table: Vec<Limb> = vec![0; ((TABLE_ENTRIES + 3) * num_limbs) + EXTRA_FOR_ALIGNMENT];
     let (table, state) = {
-        let misalignment = (table.as_ptr() as usize) % ALIGNMENT;
-        let table = &mut table[((ALIGNMENT - misalignment) / LIMB_BYTES)..];
-        assert_eq!((table.as_ptr() as usize) % ALIGNMENT, 0);
+        let align_offset = table.as_ptr().align_offset(ALIGNMENT);
+        let table = &mut table[align_offset..];
```

Comment:
Assuming the "obvious" behavior of `align_offset`, this is fine, but technically if we want to be conservative about reading this part of [the documentation](https://doc.rust-lang.org/std/primitive.pointer.html#method.align_offset):

> If it is not possible to align the pointer, the implementation returns `usize::MAX`. It is permissible for the implementation to always return `usize::MAX`. Only your algorithm’s performance can depend on getting a usable offset here, not its correctness.

...then this isn't sufficient, and we need to check for `usize::MAX`. Let me know if that's something you care about.

Reply:
> ...then this isn't sufficient, and we need to check for usize::MAX. Let me know if that's something you care about.

Well, I don't think we should rely on behavior that isn't guaranteed. In this case, it seems like there could be a more useful guarantee, and it would be good to change libcore so that such a stronger guarantee is made. But I doubt that will happen soon.

This was intended to be an intermediate step towards stack-allocating this array. I think I'm just going to skip this step and go directly to stack allocating.

Thanks for your careful review. I really appreciate it!

Reply:
Of course!